### PR TITLE
Fixed Bug. (Port In Use)

### PR DESCRIPTION
**Pull request by AlirezaP**

### CODE REVIEW

**Comment by AlirezaP** [comment]
fixed bug. (default port was in use).

**Comment by AlirezaP** [comment]
fixed bug. (default port was in use).

File: src/Microsoft.AspNetCore.Hosting/Helper/Network.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Hosting.Helper
+{
+    public class Network
+    {
+        public static bool PortValidity(int port)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Helper/Network.cs]: We have test infrastructure that does something similar except it checks by binding to the port as if it were going to be the server. I think it even tries to bind to 0 which causes it to generate and return a random port.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Helper/Network.cs]: Method name should be a verb.

File: src/Microsoft.AspNetCore.Hosting/Helper/Network.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Hosting.Helper
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Helper/Network.cs]: Hosting.Internal

File: src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs [hunk 3]
```diff
@@ -223,7 +223,22 @@ private void EnsureServer()
                 if (addresses != null && !addresses.IsReadOnly && addresses.Count == 0)
                 {
                     // Provide a default address if there aren't any configured.
-                    addresses.Add("http://localhost:5000");
+                    int port = 5000;
+
+                    //The range 49152�65535 (215+214 to 216?1) contains dynamic or private ports that cannot be registered with IANA
+                    for (int i = 49151; i < 65534; i++)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs]: This kind of loop should be in the helper method. e.g. `FindFreePort`

**Comment by Tratcher** [comment]
@DamianEdwards do we want dynamic default port selection? It makes some assumptions about the server, e.g. it may have false negatives for WebListener that can share ports. I also think the most common scenario for conflicts is when you've left a prior instance of your server running. Changing the port dynamically there could break your flow if you just went back to the browser and just hit refresh.

**Comment by davidfowl** [comment]
Not sure this should be built in. It also brings in some other dependencies...

File: src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs (hunk 3 above)
Comment (AlirezaP) [inline src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs]: yes.
you're right.

**Comment by DamianEdwards** [comment]
Yeah I don't think this should be in the box. It's also inherently prone to race-conditions leading to random failures anyway. Much prefer to just default to the single port like we do now. The idea of having this as some sort of helper isn't bad though (although it will still be racy).

**Comment by davidfowl** [comment]
Closing this out. @AlirezaP feel free to provide a different package with this functionality as an extension method on `IWebHostBuilder`

**Comment by AlirezaP** [comment]
ok
thanks

**Pull request closed without merging**